docs/pairing-architecture.md

```diff
@@ -0,0 +1,283 @@
+
+
+![](images/pairing-flow1.png)
+
+*****
+
+
+![](images/pairing-flow2.png)
+
+### Detailed Flow
+
+Is is advised to read the following section while looking at the detail digram while reading the information below.
+
+![](https://www.lucidchart.com/publicSegments/view/00b7d8a7-fd78-493f-91f5-120b4034f8b0/image.png)
+
+[View / Edit in LucidChart](https://www.lucidchart.com/documents/edit/1bc1b604-0047-4542-8827-ed8518b0433e/0)
+
+
+#### 1) Desktop opens an encrypted channel and advertises it via QR code
+
+To make itself available for pairing, the Desktop device opens a connection to the Channel Server using the [fxa-pairing-channel library](https://github.com/mozilla/fxa-pairing-channel). 
+Any application that learns the channel key will be able to request pairing from the Desktop device, or intercept messages exchanges during the pairing flow.
+This library uses a subset of TLS-like communication between the two devices.
+Once connected it will receive a server-allocated channel id, then the library will generate a 32-byte channel key to be used for client-side encryption of messages on the channel and derive a confirmation code that can be displayed to the user as visual confirmation the the devices are using the same key.
+ * The channel key is what secures the pairing flow.
+ * The channel id and channel key are encoded using base64url and then combined into a "pairing url". 
+ * The generated URL looks like: https://accounts.firefox.com/pair#channel_id=[channel_id]&channel_key=[channel_key]
+ * This URL is then encoded in a QR code and displayed on screen.
+ 
+#### 2) Mobile scans the QR code and connects to the encrypted channel
+
+To initiate pairing, the Mobile device scans the QR code displayed by Desktop and extracts the channel id and channel key.
+It opens a websocket to the Channel Server using the same fxa-pairing-channel library specifying the channel id, channel key.
+The Channel Server will now relay websocket messages from Mobile to Desktop and vice-versa.
+
+#### 3) Devices exchange metadata and show confirmation screens
+
+At this point the devices have established a baseline level of trust and can exchange metadata to allow the user to confirm the pairing on both sides. 
+
+* We use explicit confirmation screens to reduce the risk of phishing or over-the-shoulder attacks.
+
+The Mobile device sends what is essentially an OAuth request, delivered over the websocket channel rather than opened as a local webpage. 
+It includes information such as:
+```
+client_id - its registered OAuth client id
+state - OAuth state parameter
+scope - requested OAuth scopes
+code_challenge and code_challenge_method - OAuth PKCE parameters 
+keys_jwk - FxA Scoped Keys key-fetching public key
+```
+
+The Desktop device must carefully check the provided OAuth request and error out if anything seems amiss.
```

Comment:
There's more detail on these checks in the following section, maybe we can just delete this sentence.